### PR TITLE
Scopa

### PR DESCRIPTION
**Pull request by PF98**
Scopa

### CODE REVIEW

File: carte/static/games/scopa.css [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+          (var(--rotation-diff) - var(--card-radius) / 2) *
+          var(--num-offset)
+        );
+        z-index: var(--card-field-position);
+        &[data-player="opponent"] {
+          top: calc(var(--top) + var(--deck-shadow-width));
+          --card-rotation: 2; /* upside down */
+        }
+        &[data-player="self"] {
+          top: calc(var(--bottom));
```
Comment (Parnassius) [inline carte/static/games/scopa.css:201]:
```suggestion
          top: calc(var(--bottom) - var(--deck-shadow-width));
```

File: carte/static/games/scopa.css [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        font-weight: normal;
+        font-size: 0.9em;
+        line-height: 1.5em;
+
+        color: var(--background-primary-highlight-strong);
+
+        padding-left: 1em;
+        border-left: 1px solid var(--background-primary-highlight-strong);
+
+        &[data-winner] {
```
Comment (Parnassius) [inline carte/static/games/scopa.css:234]:
```suggestion
        &.winner {
```

File: carte/static/games/scopa.css [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+      &[data-player="self"] {
+        bottom: calc(var(--gap) * 2 + var(--card-height) + var(--deck-shadow-width));
+      }
+    }
+  }
+}
+
+#results {
+  #results-table {
+    tr > td[data-details] {
```
Comment (Parnassius) [inline carte/static/games/scopa.css:220]:
```suggestion
    td.details {
```

File: carte/static/games/scopa.css [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      padding-top: 0.5em;
+      padding-bottom: 0.5em;
+
```
Comment (Parnassius) [inline carte/static/games/scopa.css:223]:
```suggestion
```

File: carte/static/games/scopa.css [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      > div[data-detail-type] {
+        font-weight: normal;
+        font-size: 0.9em;
+        line-height: 1.5em;
+
```
Comment (Parnassius) [inline carte/static/games/scopa.css:228]:
```suggestion
```

File: carte/static/games/scopa.css [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        color: var(--background-primary-highlight-strong);
+
```
Comment (Parnassius) [inline carte/static/games/scopa.css:230]:
```suggestion
```

File: carte/static/games/scopa.css [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        padding-left: 1em;
+        border-left: 1px solid var(--background-primary-highlight-strong);
```
Comment (Parnassius) [inline carte/static/games/scopa.css:232]:
```suggestion
        border-left: 1px solid var(--background-secondary-highlight-strong);
```

File: carte/static/games/scopa.css [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
```
Comment (Parnassius) [inline carte/static/games/scopa.css:233]:
```suggestion
```

File: carte/static/games/scopa.css [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        &[data-winner] {
+          color: var(--text-color);
+        }
+
```
Comment (Parnassius) [inline carte/static/games/scopa.css:237]:
```suggestion
```

File: carte/static/games/scopa.css [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    tr > td[data-details] {
+      padding-top: 0.5em;
+      padding-bottom: 0.5em;
+
+      > div[data-detail-type] {
+        font-weight: normal;
+        font-size: 0.9em;
+        line-height: 1.5em;
+
+        color: var(--background-primary-highlight-strong);
```
Comment (Parnassius) [inline carte/static/games/scopa.css:229]: Please create a new variable in main.css, something like --text-color-secondary)

File: carte/static/games/scopa.css [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        padding-left: 1em;
+        border-left: 1px solid var(--background-primary-highlight-strong);
+
+        &[data-winner] {
+          color: var(--text-color);
+        }
+
+        &[data-detail-type="primiera"] > span {
+          font-family: monospace;
+          white-space-collapse: preserve-spaces;
```
Comment (Parnassius) [inline carte/static/games/scopa.css:240]:
```suggestion
```

File: carte/static/games/scopa.css [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+      }
+      &[data-starting-position="deck"] {
+        --card-starting-x: var(--left);
+        --card-starting-y: calc(var(--center-y) - var(--deck-shadow-width) / 2);
+      }
+
+      &[data-position="playing-area"] {
+        --selected-card-offset: 0;
+
+        /* is 0 for a --card-field-position in the last row, 1 for any previous row */
```
Comment (Parnassius) [inline carte/static/games/scopa.css:108]: Isn't it -1 for the non-last rows?

**Review by Parnassius — CHANGES_REQUESTED** [review]

File: carte/static/games/base.css [hunk 13]
```diff
@@ -212,6 +212,10 @@
         &:last-child {
           padding-right: 1em;
         }
+
+        &[data-points] {
+          text-align: right;
+        }
```
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/base.css:218]: ok

File: carte/static/games/base.css [hunk 14]
```diff
@@ -88,7 +88,7 @@
         &[data-deck-count="0"] {
           visibility: hidden;
         }
-        &[data-deck-count]:hover {
+        &:not([data-deck-capped])[data-deck-count]:hover {
```
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/base.css:91]: ok

File: carte/static/games/base.js [hunk 15]
```diff
@@ -101,19 +106,25 @@ class Deck extends CardGroup {
       deckParams.set(field, this.params.get(field));
     }
 
+    if (this.capped) {
+      deckParams.set("deckCapped", this.capped);
```
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/base.js:110]: ok

File: carte/static/games/base.js [hunk 16]
```diff
@@ -684,6 +703,7 @@ class BaseGame {
       const playerCell = row.insertCell();
       playerCell.textContent = this.players[playerId];
       const pointsCell = row.insertCell();
+      pointsCell.dataset.points = "";
```
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/base.js:706]: ok

File: carte/static/games/scopa.css [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    /* TAKEABLE CARDS */
+    &[data-playing] .card[data-position="playing-area"] {
+      &[data-takeable],
+      &[data-selected] {
+        &::after {
+          display: block;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/scopa.css:20]: ok, the :is(...) syntax is quite convenient

File: carte/static/games/scopa.css [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+      --unit: min(
+        calc(
+          100dvw /
+          (
+            var(--card-height-raw) +
+            var(--card-width-raw) *
+            8 +
+            var(--gap-raw) *
+            10 +
+            var(--deck-shadow-max-width-raw)
```
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/scopa.css:72]: oops

File: carte/static/games/scopa.css (hunk 1 above)
Comment (PF98) [reply carte/static/games/scopa.css:201]: oof

File: carte/static/games/scopa.css (hunk 2 above)
Comment (PF98) [reply carte/static/games/scopa.css:234]: ok

File: carte/static/games/scopa.css (hunk 3 above)
Comment (PF98) [reply carte/static/games/scopa.css:220]: ok

File: carte/static/games/scopa.css (hunk 7 above)
Comment (PF98) [reply carte/static/games/scopa.css:232]: ok

File: carte/static/games/scopa.css (hunk 10 above)
Comment (PF98) [reply carte/static/games/scopa.css:229]: ok

File: carte/static/games/scopa.css (hunk 11 above)
Comment (PF98) [reply carte/static/games/scopa.css:240]: oops

File: carte/static/games/scopa.css (hunk 12 above)
Comment (PF98) [reply carte/static/games/scopa.css:108]: whoopsie daisy, yes

File: carte/static/games/scopa.css (hunk 18 above)
[the remark this answers is not in the document]
Comment (PF98) [reply carte/static/games/scopa.css]: done

File: carte/games/scopa.py [hunk 19]
```diff
@@ -0,0 +1,443 @@
...
+                if self._table:
+                    await self._send("take_all", self._last_taker_id)
+                    self._players[self._last_taker_id].points.extend(self._table)
+                    self._table.clear()
+
+                self._game_status = GameStatus.ENDED
+
+                await self._send("results_prepare")
+
+                results = [0 for _ in self._players]
```
Comment (Parnassius) [inline carte/games/scopa.py:272]:
```suggestion
```

**Review by Parnassius — CHANGES_REQUESTED** [review]

File: carte/games/scopa.py [hunk 20]
```diff
@@ -0,0 +1,443 @@
...
+    async def _draw_full_hands(self) -> None:
+        for _ in range(self.hand_size):
+            for i in range(self.number_of_players):
+                player_id = (self._current_player_id + i) % self.number_of_players
+                await self._draw_card(self._players[player_id])
+
+    @cmd(game_status=GameStatus.STARTED, current_player=True)
+    async def cmd_play(self, card: Card) -> None:
+        # wrong playing status
+        if self._playing_status != ScopaPlayingStatus.HAND:
```
Comment (Parnassius) [inline carte/games/scopa.py:140]:
```suggestion
        if self._playing_status is not ScopaPlayingStatus.HAND:
```

File: carte/games/scopa.py [hunk 21]
```diff
@@ -0,0 +1,443 @@
...
+                return
+
+        await self._send("turn_status", self._playing_status)
+
+        await self._send(self.current_player, "turn")
+
+    @cmd(game_status=GameStatus.STARTED, current_player=True)
+    async def cmd_take_choice(self, card: Card) -> None:
+        # wrong playing status
+        if self._playing_status != ScopaPlayingStatus.CAPTURE:
```
Comment (Parnassius) [inline carte/games/scopa.py:184]:
```suggestion
        if self._playing_status is not ScopaPlayingStatus.CAPTURE:
```

**Review by Parnassius — CHANGES_REQUESTED** [review]

**Pull request merged**